rfcs/20191212-keras-categorical-inputs.md

```diff
@@ -0,0 +1,390 @@
+
+Original:
+```python
+fc = tf.feature_column.categorical_column_with_hash_bucket(
+       key, hash_bucket_size, dtype)
+```
+Proposed:
+```python
+x = tf.keras.Input(shape=(1,), name=key, dtype=dtype)
+layer = tf.keras.layers.CategoryHashing(num_bins=hash_bucket_size)
+out = layer(x)
+```
+
+4. categorical identity column
+
+Original:
+```python
+fc = tf.feature_column.categorical_column_with_identity(
+       key, num_buckets, default_value)
+```
+Proposed:
+```python
+x = tf.keras.Input(shape=(1,), name=key, dtype=dtype)
+layer = tf.keras.layers.Lambda(lambda x: tf.where(tf.logical_or(x < 0, x > num_buckets), tf.fill(dims=tf.shape(x), value=default_value), x))
+out = layer(x)
+```
+
+5. cross column
+
+Original:
+```python
+fc_1 = tf.feature_column.categorical_column_with_vocabulary_list(key_1, vocabulary_list, 
+         dtype, default_value, num_oov_buckets)
+fc_2 = tf.feature_column.categorical_column_with_hash_bucket(key_2, hash_bucket_size,
+         dtype)
+fc = tf.feature_column.crossed_column([fc_1, fc_2], hash_bucket_size, hash_key)
+```
+Proposed:
+```python
+x1 = tf.keras.Input(shape=(1,), name=key_1, dtype=dtype)
+x2 = tf.keras.Input(shape=(1,), name=key_2, dtype=dtype)
+layer1 = tf.keras.layers.CategoryLookup(
+           vocabulary=vocabulary_list,  
+           num_oov_tokens=num_oov_buckets)
+x1 = layer1(x1)
+layer2 = tf.keras.layers.CategoryHashing(
+           num_bins=hash_bucket_size)
+x2 = layer2(x2)
+layer = tf.keras.layers.CategoryCrossing(num_bins=hash_bucket_size)
+out = layer([x1, x2])
+```
+
+6. weighted categorical column
+
+Original:
+```python
+fc = tf.feature_column.categorical_column_with_vocab_list(key, vocabulary_list,
+         dtype, default_value, num_oov_buckets)
+weight_fc = tf.feature_column.weighted_categorical_column(fc, weight_feature_key, 
+         dtype=weight_dtype)
+linear_model = tf.estimator.LinearClassifier(units, feature_columns=[weight_fc])
+```
+Proposed:
+```python
+x1 = tf.keras.Input(shape=(1,), name=key, dtype=dtype)
+x2 = tf.keras.Input(shape=(1,), name=weight_feature_key, dtype=weight_dtype)
+layer = tf.keras.layers.CategoryLookup(
+           vocabulary=vocabulary_list,   
+           num_oov_tokens=num_oov_buckets)
+x1 = layer(x1)
+x = tf.keras.layers.CategoryEncoding(num_categories=len(vocabulary_list)+num_oov_buckets)([x1, x2])
+linear_model = tf.keras.premade.LinearModel(units)
+linear_logits = linear_model(x)
+```
+
+## Design Proposal
+We propose a CategoryLookup layer to replace `tf.feature_column.categorical_column_with_vocabulary_list` and `tf.feature_column.categorical_column_with_vocabulary_file`, a `CategoryHashing` layer to replace `tf.feature_column.categorical_column_with_hash_bucket`, a `CategoryCrossing` layer to replace `tf.feature_column.crossed_column`, and another `CategoryEncoding` layer to convert the sparse input to the format required by linear models.
+
+```python
+`tf.keras.layers.CategoryLookup`
+CategoryLookup(PreprocessingLayer):
```

Comment:
Please don't forget to implement the correct `compute_output_signature` for these classes, since they will accept SparseTensorSpecs, and must emit SparseTensorSpecs in this case.

Reply:
Thanks for the reminder!